Update to qemu 3.0.0 and security enhancements

alcreator opened the pull request:
Qemu 3.0.0 incorporates some fixes for Xen PCI passthrough.

Summary of Changes:
-Update to qemu 3.0.0.
-Modify init script to use inotify instead of polling (reduces idle CPU usage, https://github.com/QubesOS/qubes-issues/issues/2849).
-Handle ifup/ifdown scripts externally from qemu, and set seccomp spawn filter to deny.
-Remount rootfs to readonly.
-Build a hardened busybox (as udhcpd is used, and as busybox versions included in fedora 27/28 appear to have some issues with the init script, this will simplify future upgrades).

Requires https://github.com/QubesOS/qubes-vmm-xen/pull/47 and https://github.com/QubesOS/qubes-gui-agent-xen-hvm-stubdom/pull/5

alcreator committed:
Update to qemu 3.0.0

alcreator committed:
Update init script to use inotify instead of polling

alcreator committed:
Update seccomp filter to deny spawning, and remount rootfs to read only

alcreator committed:
Build hardened busybox

marmarek commented:
How big is resulting busybox binary? Can we save some memory (it's all in initramfs) by building smaller one?

alcreator commented:
It's a similar size to the original one, it should be possible to reduce the size of it by omitting unnecessary applets from the config file (I copied the one from fedora 29 srcrpm). It's only approx 1mb large, so it won't make much difference.

marmarek commented:
offtopic: any other ideas to reduce memory usage for stubdom?

alcreator commented:
Changing the stubdom rootfs use a block device instead of being loaded in domain ram should save ~25mb (the size of the uncompressed rootfs).

alcreator commented:
Changing the libc version to uClibc-ng or musl might also reduce memory usage. I have previously tested a build using 32-bit musl (build using Alpine linux), although qemu performance was reduced, and it had slightly higher memory usage. It may have been due to the rootfs size, I can't remember if it was bigger than the standard build. Another consideration is that the compiler hardening options may not be as mature as the glibc implementations.

marmarek commented:
> Changing the stubdom rootfs use a block device instead of being loaded in domain ram should save ~25mb (the size of the uncompressed rootfs).

We had it before, but changed to initramfs for simplicity and to avoid potentially fatal mistake of setting rw dev here. Most of it will end up in RAM to be executed anyway (if not, IMHO shouldn't be included at all). Q: is binary on initramfs executed directly from there, or copied in RAM?

alcreator commented:
According to [https://stackoverflow.com/questions/25270110/is-executable-code-stored-uniquely-in-tmpfs-copied-to-another-part-of-ram-when-r](url), the executable's pages do get directly mapped without copying, however I don't know if this is still the case for executables compiled with PIE/ASLR.

I know that some filesystem types (such as cramfs - see [https://github.com/torvalds/linux/blob/master/Documentation/filesystems/cramfs.txt](url)) allow for XIP support, so if ASLR programs can still be loaded in this way, it may still save ram to have it on a block device.

marmarek commented:
> if this is still the case for executables compiled with PIE/ASLR.

Yes, ELF structure for PIE binaries require only very few pages to be modified to load code at arbitrary address (namely, global offset table). The actual code pages remains unchanged, so can be mapped directly (if underlying mechanism support it, but the above says so).

marmarek commented:
Could you update submodule ref, so the build will work (I've merged gui-agent PR already)?

alcreator committed:
Update submodule reference

alcreator committed:
Update busybox version

alcreator committed:
mount tmpfs with nodev,noexec and add additional seccomp call filters

alcreator committed:
add delay to qemu-ifdown script

alcreator commented:
I've just updated the busybox version to 1.29.3, and added some additional filters to the seccomp call list for qemu.
https://github.com/QubesOS/qubes-issues/issues/4191 Might be caused by the vif adapter being removed too quickly from xenstore, so I've added a delay to the ifdown script when it is called. I don't know if this will definitely fix the problem, since I can't consistently reproduce this issue.

marmarek commented:
I wonder if that xenstore-rm is the right thing to do... The purpose of xenstore-rm is to close the interface, so the backend will know that PV backend device the the actual domain (not stubdomain) should be used (see [vif-route-qubes](https://github.com/QubesOS/qubes-core-agent-linux/blob/master/network/vif-route-qubes#L91)).
But generally, xenstore entries are owned by toolstack and drivers, so modifying them from frontend's userspace is a fragile thing. The `xenstore-rm` is similar to what `xl -f network-detach` does. Unfortunately, just `xl network-detach` writes to backend's xenstore entries, so it's out of reach for the frontend domain. But maybe the same could be achieved using something else? Some sysfs write?

Anyway, this is independent of qemu update here.

rustybird commented on rootfs/init:
```diff
@@ -92,7 +92,8 @@ echo "$line" >&2
     if [ "$id" = "" ]; then
-        sleep 0.1
+        sleep 0.1; if [ -n "$(ls -t /tmp/qmp/req)" ]; then touch /tmp/qmp/req; fi &
+        inotifywait -e attrib,create,delete,modify,move /tmp/qmp/req > /dev/null 2>&1
```
This looks like a race condition on every loop iteration. (**Edit:** And parentheses are missing around `sleep ... fi`.) Probably hard to say how long it's going to take `inotifywait` to start and set up the inotify watches, e.g. during high system load?

`inotifywait` doesn't make it easy to _first_ set up the watches and _then_ check if anything interesting happened in the meantime. The best I can think of is something ugly like

```
inotifywait -m -e moved_to --format %f /tmp/qmp/req 2>&1 |
grep --line-buffered . | {
    read _  # "Setting up watches."
    read _  # "Watches established."
    if id=$(ls -t /tmp/qmp/req | grep -m 1 .); then
        # handle "$id"
    fi

    while read -r id; do
        # handle "$id"
    done
}
```

But I haven't really looked at the script, maybe there's a better way. Or maybe the race doesn't even matter in practice.

alcreator replied:
Yes, I thought this might result in a race condition too, that why I originally added `sleep 0.1; if [ -n "$(ls -t /tmp/qmp/req)" ]; then touch /tmp/qmp/req; fi` (which definitely should be in parentheses), but I think your method is a better way of handling it.

Something like this can be slotted straight into the script:

```
inotifywait -e attrib,create,delete,modify,move /tmp/qmp/req 2>&1 |
grep --line-buffered . | {
    while read -r watch_setup; do
        if [ "$watch_setup" = "Watches established." ]; then
           if [ -n "$(ls -t /tmp/qmp/req)" ]; then
               touch /tmp/qmp/req
           fi
        fi
    done
}
```
I'll keep the `create,delete,modify` watches incase the qmp code is extended in the future.

alcreator committed:
Fix potential inotifywait race condition

alcreator committed:
Add upstream seccomp patches

marmarek commented:
Qemu prints `qemu: pin_memory_cacheattr failed: Operation not permitted`. Not sure if that's harmful.

alcreator commented:
Qemu commit https://github.com/qemu/qemu/commit/2cbf8903530b936964dd3af7e2e5bf85c3955d5c introduces an error check for the `xendevicemodel_pin_memory_cacheattr` function. On previous versions of qemu, this call silently fails, at least when running in a stubdomain. So this isn't a new issue and dosen't seem to cause any problems with the emulated VGA.

On the subject of reducing memory usage, it seems that using a block device instead of initramfs does actually save some memory:

Regular initramfs (32-bit):
````

# mount
rootfs on / type rootfs (ro)
none on /dev type devtmpfs (rw,relatime)
/sys on /sys type sysfs (rw,relatime)
/proc on /proc type proc (rw,relatime)
/tmp on /tmp type tmpfs (rw,nodev,noexec,relatime)
/dev/xvdd on /tmp/boot type ext3 (ro,relatime,data=ordered)
# free
             total       used       free     shared    buffers     cached
Mem:        138704      75968      62736          0        152      35812
-/+ buffers/cache:      40004      98700
Swap:            0          0          0
# echo 3 > /proc/sys/vm/drop_caches 
sh (856): drop_caches: 3
# free
             total       used       free     shared    buffers     cached
Mem:        138704      63088      75616          0          8      23068
-/+ buffers/cache:      40012      98692
Swap:            0          0          0
````

External Rootfs:
Kernel command line: debug console=hvc0 root=/dev/xvdi rootfstype=ext2 init=/init
````

# mount
/dev/root on / type ext2 (ro,relatime,block_validity,barrier,user_xattr)
none on /dev type devtmpfs (rw,relatime)
/sys on /sys type sysfs (rw,relatime)
/proc on /proc type proc (rw,relatime)
/tmp on /tmp type tmpfs (rw,nodev,noexec,relatime)
/dev/xvdd on /tmp/boot type ext3 (ro,relatime,data=ordered)
# free
             total       used       free     shared    buffers     cached
Mem:        138720      74768      63952          0        456      29808
-/+ buffers/cache:      44504      94216
Swap:            0          0          0
# echo 3 > /proc/sys/vm/drop_caches 
sh (846): drop_caches: 3
# free
             total       used       free     shared    buffers     cached
Mem:        138720      55700      83020          0        104      11076
-/+ buffers/cache:      44520      94200
Swap:            0          0          0
````

marmarek commented:
Hmm, I get `Bad ram offset 806f5000` (and qemu crash) from time to time when running Windows 7, even before this PR. Maybe it is related?

> On the subject of reducing memory usage, it seems that using a block device instead of initramfs does actually save some memory:

I assume both are with running qemu and a VM which does something (not completely idle), right? I wonder how fast it loads this 8MB back to RAM...

marmarek commented:
Potential problem with using block dev for stubdom root - convince libxl to attach a disk to a stubdomain, but not target domain. And avoid conflicts with target domain's disks. The last one is trivial in Qubes case, but not necessary in generic case (which is important, as we're [in a process of upstreaming this part](https://markmail.org/thread/l3wazwte53ob7bpi)).

alcreator commented:
> Hmm, I get Bad ram offset 806f5000 (and qemu crash) from time to time when running Windows 7, even before this PR. Maybe it is related?

That would most likely be https://github.com/QubesOS/qubes-issues/issues/3651. Maybe the memory pinning could have something to do with it, but I think that issue is something related to xen-mapcache.

> I assume both are with running qemu and a VM which does something (not completely idle), right? I wonder how fast it loads this 8MB back to RAM...

This was with a standard fedora-28 hvm (based off qubes template). So it is using the pv disk and network drivers. I haven't verified this, but I think some unused parts of the libraries can get dropped from the cache. Without pv drivers, I'd imagine that the caches will fill up relatively quickly with any disk i/o, but in many cases, cache memory can be reclaimed by the kernel. So this was more to show that the vm has more reclaimable memory, compared to the initramfs rootfs.

Compiling qemu and busybox with -Os, and running strip --strip-unneeded on the /lib directory saves ~2MB from the rootfs. Recompiling the libraries with -Os, or static linking with uClibc-ng/musl may reduce this more. The kernel itself also seems to consume over 20mb ram, which is still fairly major compared to the savings obtained with a smaller rootfs, and probably can't be avoided without switching to a different microkernel.
